[8.0] product visible discount extension

veloutin opened the pull request.

llacroix committed:
[#64106] Does the changes correctly

llacroix committed:
[#64106] Fix PEP8 errors and changed the formula to calculate correctly the price_unit

llacroix committed:
[#64106] Condition isn't needed as we delete from 100.

llacroix committed:
[#64106] Set the discount to 0 and use just a visual discount field

llacroix committed:
Override product_id_change to setup the model correctly

llacroix committed:
Fix PEP8 errors

llacroix committed:
Added ValidationError when visible_discount > 100

llacroix committed:
Added the visible discount to the sale report

llacroix committed:
Added visible discount to report and changed the discount label

llacroix committed:
Adding support for invoice report

llacroix committed:
Added account_invoice.py to add visible_discount to account_invoice_line

llacroix committed:
Added the visible_discount to the form view

llacroix committed:
Move file in views/models folders

llacroix committed:
Fix PEP8 errors

llacroix committed:
Fixed rounding errors. Numbers are rounded before sending them

It prevent recalculating the values entered.

llacroix committed:
Copy the visible discount when creating invoices from delivery order

llacroix committed:
Fix price_unit getting inside visible_discount

pedrobaeza commented on product_visible_discount_extension/__openerp__.py:
```diff
@@ -0,0 +1,57 @@
+    'category': 'Others',
+    'summary': 'Product visible discount extension',
+    'description': """
+Product visible discount extension
```
Please extract it to a README.rst file.

pedrobaeza commented on product_visible_discount_extension/__openerp__.py:
```diff
@@ -0,0 +1,57 @@
+Product visible discount extension
+==================================
+
+This module is based on product_visible_discount that calculate
```
Better to put "This module extends product_visible_discount..."

pedrobaeza commented on product_visible_discount_extension/__openerp__.py:
```diff
@@ -0,0 +1,57 @@
+the discount rate on unit price calculated from a price list in
+sale order lines. The discount rate can be either calculated from
+the Public Price of the product and the unit price written in
+the sale order line or set by the user.
```
It is not clear what this module extends exactly.

pedrobaeza commented on product_visible_discount_extension/models/sale.py:
```diff
@@ -0,0 +1,189 @@
+                              related="product_id.list_price",
+                              store=True,
+                              readonly=True)
+    visible_discount = fields.Float('Customer  Discount (%)')
```
Why are you using a new field instead of the existing one?

llacroix replied:
There is a problem, if I use the real discount, it modify the `price_unit` value to the actual price it would get with the discount calculated.. Which means we can't calculate the discount again. If we were using the discount field, we would have something like this:

```
price_unit = (list_price * (1 - discount))
discount = discount
total price = price_unit * (1 - discount)
```

The result would be to calculate twice the same discount. The problem comes from openerp not showing the price after discount.

One other way to do, would be to use the price_unit as list_price without touching it. Calculating the discount and having an other field to calculate the price after discount. In the end, the discount field is used as usual, the price_unit would be exactly the same as list_price. The price after discount would have no effect anywhere else. 

But we would want to show the price after discount in reports.

pedrobaeza replied:
OK, get it, but I see no relation then with product_visible_discount module.

llacroix replied:
The product visible discount does the changes on the product with the price list. When "visible discount" isn't checked in the price lists, price lists don't seem to have any effect. Yet the module should be working even without that if I'm not mistaken I can check if the dependencies are really necessary.

llacroix committed:
Fix error on visible_discount not correctly set

llacroix committed:
Added tests (don't run onchange event)

llacroix committed:
Extended test

llacroix committed:
Forgot debug break point

llacroix committed:
Reset order line before tests

As tests are apparently directly saving data to the database.
We have to reset the order line to default values. So each tests
are expecting the same initial values.

llacroix commented:
Tests are passing on this module, there is a module that doesn't seem to work. I believe it's related to a test creating a model without defining a view name. I'm not sure if it's a bug in odoo or There is something specific. I had the same problem with my module. For some reasons it required me to use something like this:

```
!record {model..., id... view=False}:
```

llacroix commented:
If someone has an idea how to trigger change events from within the tests I'll update it that way. It's not super clean this way. Since some fields are mutual, it should trigger recursive on change events. I'm currently triggering them manually so it should always work that way. But it would be nice to see have odoo handle the event by itself and check that no matter how the events are triggered it will work.

llacroix committed:
Tests in the wrong section

llacroix committed:
Use individual fields to keep track of changes.

The order in which events are going to get triggered is not
determinable. On some databases, it will trigger the onchange for
prices and on some other databases it will trigger the onchange on
discount first.

It might depend on order that the modules are loaded.

llacroix committed:
Fix PEP8 errors

llacroix committed:
Changed label Customer Discount to Discount and translations

llacroix commented:
Some of the bugs in travis are related to this issue as far as I can say:
https://github.com/odoo/odoo/pull/5410

A view name seems to be missing  on an inline view and the code that is supposed to log a debug messages raise a KeyError.

llacroix committed:
fields with store=False require a search function

llacroix committed:
PEP8 errors

llacroix committed:
Fix tests

llacroix committed:
PEP8 errors

llacroix committed:
Added fix for tests

joaoalf commented on product_visible_discount_extension/models/sale.py:
```diff
@@ -0,0 +1,206 @@
+        _onchange_visible_discount
+    """
+
+    _name = 'sale.order.line'
```
When you use _inherit is not necessary to add _name unless you want to create a different model.

llacroix replied:
Right.

joaoalf commented on product_visible_discount_extension/models/sale.py:
```diff
@@ -0,0 +1,206 @@
+)
+
+
+def dummy_search(self, *a, **ka):
```
@llacroix  I'm not sure that this is the way to go. Can you clarify a little the usage of temp_discount and tempo_price?

llacroix replied:
As there is no real way to silence change events and that these onchange events are mutual. A change to field A will trigger a change to field B that will trigger a change to field A again. This isn't probably the best way to handle this.

joaoalf commented on product_visible_discount_extension/models/sale.py:
```diff
@@ -0,0 +1,206 @@
+                # a discount
+                product_obj = self.pool.get('product.product')
+                product = product_obj.browse(cr, uid, product, context)
+                price_list = product.list_price
```
@llacroix Are you considering the pricelist in the SO or just the product's base price?

llacroix replied:
I'm only considering the base price which is stored in the pricelist. The price_unit is correctly set to the price defined in the price_list and then we can modify the price freely. The price or discount set is the actual price that will be used later.

llacroix committed:
Added docstring to sale.py explaining what should be improved in the
module

llacroix committed:
Point out the file that needs to be fixed/improved

gurneyalex commented on product_visible_discount_extension/__openerp__.py:
```diff
@@ -0,0 +1,62 @@
+{
+    'name': 'Product visible discount extension',
+    'version': '0.1',
+    'author': 'Savoir-faire Linux',
```
please add Odoo Community Association (OCA) to authors list.

gurneyalex commented:
@llacroix any update on this PR?

pedrobaeza commented:
As there's no answer, I close. Feel free to reopen in case there's interest.
